### PR TITLE
[WD-12166] feat: Added Login/Logout button (ui) for LXD-SM.

### PR DESCRIPTION
**Pull request by Kxiru**
Tasks:
* Added a route to the Login Page
* Added a Login page using LoginPageLayout
* Added a Login button to the login page and a logout button to the SiteList page.

Login Page
image

Logout Button
image

### CODE REVIEW

File: ui/src/App.tsx [hunk 1]
```diff
@@ -2,12 +2,14 @@ import { FC, lazy, Suspense } from "react";
...
 
 const SiteList = lazy(() => import("pages/sites/SiteList"));
+const Login = lazy(() => import("pages/Login"));
 
 const App: FC = () => {
   return (
     <Suspense fallback={<div>Loading</div>}>
       <Routes>
         <Route path="/" element={<Navigate to="/1.0/ui" replace={true} />} />
+        <Route path="/ui/login" element={<Login />} />
```
Comment (edlerd) [inline ui/src/App.tsx]:
```suggestion
        <Route path="/ui" element={<Login />} />
```
I think this should be the default route, the one below can be /ui/sites

File: ui/src/pages/Login.tsx [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import { FC } from "react";
+import { Button, LoginPageLayout } from "@canonical/react-components";
+import { Link } from "react-router-dom";
+
+const Login: FC = () => {
+  return (
+    <LoginPageLayout title="Login">
+      <Button>
+        <Link to={"/1.0/oidc/login"}>Login (OIDC)</Link>
```
Comment (edlerd) [inline ui/src/pages/Login.tsx]: There shouldn't be a link inside a button, you can set the href param on the button instead. The caption should be "Login", no need to point out we are using oidc to the user.

File: ui/src/pages/Login.tsx [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import { FC } from "react";
+import { Button, LoginPageLayout } from "@canonical/react-components";
+import { Link } from "react-router-dom";
+
+const Login: FC = () => {
+  return (
+    <LoginPageLayout title="Login">
```
Comment (edlerd) [inline ui/src/pages/Login.tsx]:
```suggestion
    <LoginPageLayout title="Login to LXD site manager">
```
Maybe better?

File: ui/src/pages/sites/SiteList.tsx [hunk 4]
```diff
@@ -29,6 +30,10 @@ const SiteList: FC = () => {
           };
         })}
       />
+
+      <Button>
+        <Link to={"/1.0/oidc/logout"}>Login (OIDC)</Link>
```
Comment (edlerd) [inline ui/src/pages/sites/SiteList.tsx]:
```suggestion
        <Link to="/1.0/oidc/logout">Logout</Link>
```

**Review by edlerd — CHANGES_REQUESTED** [review]

File: ui/src/pages/Login.tsx (hunk 2 above)
Comment (mas-who) [reply ui/src/pages/Login.tsx]: Please also change the link url to `/oidc/login` instead, we planning on getting rid of the `/1.0` prefix for login endpoints

File: ui/src/pages/sites/SiteList.tsx (hunk 4 above)
Comment (mas-who) [inline ui/src/pages/sites/SiteList.tsx]:
```suggestion
        <Link to={"/oidc/logout"}>Login (OIDC)</Link>
```

**Review by edlerd — CHANGES_REQUESTED** [review]
Code and QA LGTM, can you ensure the commit message is reflecting the changes?

For the future, you don't need to use the project name in the branch name, as the branches are namespaced per project.

**Comment by edlerd** [comment]
You can use `git commit --amend -a -S -s` to change the commit message.

**Review by edlerd — APPROVED** [review]
All good now, thanks for adding it and providing feedback to improve the setup / readme when starting.

Just one issue with the failing e2e test, I think we need to update the test with the new home page title.

**Comment by Kxiru** [comment]
> All good now, thanks for adding it and providing feedback to improve the setup / readme when starting.
> 
> Just one issue with the failing e2e test, I think we need to update the test with the new home page title.

All fixed!

**Comment by Kxiru** [comment]
Although I believe the Chromium e2e test is failing incorrectly. As the information is it pulling is outdated. The title of the Login.tsx page has definitely been changed, and it is aligned with the correct navigation url.

**Comment by edlerd** [comment]
> Although I believe the Chromium e2e test is failing incorrectly. As the information is it pulling is outdated. The title of the Login.tsx page has definitely been changed, and it is aligned with the correct navigation url.

This is probably a timing issue. Because the login page is changing the title after the initial loading. You could also change the test to rely on the /ui/sites page to have the previous title. After all that is the site list spec test.

**Comment by Kxiru** [comment]
@edlerd After the alterations we made yesterday, the tests are still failing. How would you like me to approach this?

**Comment by edlerd** [comment]
> @edlerd After the alterations we made yesterday, the tests are still failing. How would you like me to approach this?

I think we are missing the latest changes from the main branch and that's why the request to /ui/sites is 404. Can you rebase on main and force-push the update? That should fix it.

**Comment by Kxiru** [comment]
I have already rebased and forced pushed here :(. Which changes are missing?

**Comment by edlerd** [comment]
> I have already rebased and forced pushed here :(. Which changes are missing?

You are right indeed. I think this is because /ui/sites is a new path that needs to be registered in go explicitly. @mas-who can we make it like in LXD where all paths under /ui/* serve the html/js frontend files? If not, can you help to register the new path in go?

**Comment by mas-who** [comment]
> > I have already rebased and forced pushed here :(. Which changes are missing?
> 
> You are right indeed. I think this is because /ui/sites is a new path that needs to be registered in go explicitly. @mas-who can we make it like in LXD where all paths under /ui/* serve the html/js frontend files? If not, can you help to register the new path in go?

I've created a PR #22 to resolve the issue here. Left some comments on that PR for context and I think there is still a bit of work to be done to streamline how the ui should be served.

**Comment by Kxiru** [comment]
@edlerd , @mas-who , thanks for making the fixing PR! What does this mean for this PR though? :)

**Comment by mas-who** [comment]
> @edlerd , @mas-who , thanks for making the fixing PR! What does this mean for this PR though? :)

The other PR will need to be merged first, then the failing CI here should pass since I've tested the changes with your PR.

**Comment by Kxiru** [comment]
> > @edlerd , @mas-who , thanks for making the fixing PR! What does this mean for this PR though? :)
> 
> The other PR will need to be merged first, then the failing CI here should pass since I've tested the changes with your PR.

As vigilant as ever. Thanks a bunch!

**Comment by mas-who** [comment]
@Kxiru the #22 just got merged. Can you rebase the changes and push again

**Pull request merged**